crates/noirc_frontend/src/hir/def_collector/dc_crate.rs

```diff
@@ -220,6 +245,107 @@ fn collect_impls(
     }
 }
 
+fn collect_global_constants(
+    context: &mut Context,
+    global_constants: Vec<UnresolvedGlobalConst>,
+    crate_id: CrateId,
+    errors: &mut Vec<CollectedErrors>,
+) {
+    for global_constant in global_constants {
+        let path_resolver = StandardPathResolver::new(ModuleId {
+            local_id: global_constant.module_id,
+            krate: crate_id,
+        });
+
+        let mut resolver = Resolver::new(
+            &mut context.def_interner,
+            &path_resolver,
+            &context.def_maps,
+            global_constant.file_id,
+        );
+
+        let name = match global_constant.stmt_def.clone() {
+            Statement::Let(let_stmt) => {
+                match let_stmt.pattern {
+                    Pattern::Identifier(ident) => ident,
+                    _ => panic!("pattern for const statement must be an identifier"), // TODO: change this to use errors
```

Comment:
I see what you mean here, and I was considering how to change this up as well as I didn't like this check here. Originally I was thinking to add a method like `extract_stmt_name` to the resolver for an ast stmt. Or make a resolution specifically for let stmts. By "swap out the scoping part where the name is defined" do you mean changing how `resolve_pattern` operates so that we do not define a name in `add_variable_decl`? Is this what you mean by patterns for "free"?

Reply:
I added a trait method to Pattern to easily fetch the identifier name without this special check. I also changed how we collect global constants that I think is much cleaner and only defines the name on resolution